include/seqan3/alignment/matrix/alignment_trace_matrix.hpp

```diff
@@ -0,0 +1,333 @@
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \author Marcel Ehrhardt <marcel.ehrhardt AT fu-berlin.de>
+ * \brief Contains the declaration of seqan3::detail::alignment_trace_matrix.
+ */
+
+#pragma once
+
+#include <list>
+#include <seqan3/alphabet/gap/gapped.hpp>
+
+#include <seqan3/alignment/matrix/matrix_concept.hpp>
+#include <seqan3/alignment/matrix/alignment_score_matrix.hpp>
+#include <seqan3/core/add_enum_bitwise_operators.hpp>
+#include <seqan3/core/metafunction/basic.hpp>
+
+namespace seqan3::detail
+{
+
+/*!\brief The possible directions a trace can have. The values can be combined by the logical `|`-operator.
+ * \ingroup alignment_matrix
+ * \sa seqan3::add_enum_bitwise_operators <seqan3::detail::trace_directions> enables combining enum values.
+ * \sa seqan3::detail::alignment_trace_matrix implementations use this enum as matrix entry type.
+ */
+enum struct trace_directions : uint8_t
+{
+    //!\brief No trace
+    none      = 0b0000,
+    //!\brief Trace comes from the diagonal entry.
+    diagonal  = 0b0001,
+    //!\brief Trace comes from the above entry.
+    up        = 0b0010,
+    //!\brief Trace comes from the left entry.
+    left      = 0b0100
+};
+
+} // namespace seqan3::detail
+
+namespace seqan3
+{
+//!\brief Enable bitwise operators for enum seqan3::detail::trace_directions.
+//!\ingroup alignment_matrix
+template <>
+constexpr bool add_enum_bitwise_operators<seqan3::detail::trace_directions> = true;
+} // namespace seqan3
+
+namespace seqan3::detail
+{
+
+//!\brief The declaration of alignment_trace_matrix. Each definition of this
+//!       declaration must satisfy seqan3::detail::matrix_concept.
+//!\ingroup alignment_matrix
+//!\implements seqan3::detail::matrix_concept
+template <typename ...>
+struct alignment_trace_matrix;
+
+/*!\brief Compute the trace from a trace matrix
+ * \ingroup alignment_matrix
+ * \tparam    database_t     The type of the database sequence.
+ * \tparam    query_t        The type of the query sequence.
+ * \tparam    trace_matrix_t The type of the trace matrix.
+ * \param[in] database       The database sequence.
+ * \param[in] query          The query sequence.
+ * \param[in] matrix         The trace matrix.
+ */
+template <
+    typename database_t,
+    typename query_t,
+    typename trace_matrix_t,
+    typename gapped_database_t = gapped<std::decay_t<decltype(std::declval<database_t>()[0])>>,
+    typename gapped_query_t = gapped<std::decay_t<decltype(std::declval<query_t>()[0])>>>
+//!\cond
+    requires matrix_concept<remove_cvref_t<trace_matrix_t>> &&
+             std::Same<typename remove_cvref_t<trace_matrix_t>::entry_type, trace_directions>
+//!\endcond
+inline std::pair<std::list<gapped_database_t>, std::list<gapped_query_t>>
```

Comment:
this returns a pair of lists of sequences? shouldn't it be pairs of numbers?

Reply:
ah, ok, so `gapped_database_t` is actually `gapped_database_alphabet_t`? Could we rename it? It's very confusing.